engine/src/test/java/io/zeebe/engine/util/client/JobClient.java

```diff
@@ -49,6 +49,7 @@ public JobClient(final StreamProcessorRule environmentRule) {
 
   public JobClient ofInstance(final long workflowInstanceKey) {
     this.workflowInstanceKey = workflowInstanceKey;
+    jobRecord.setWorkflowInstanceKey(workflowInstanceKey);
```

Comment:
This was necessary to fix one of the tests. The test used this method to drive the engine. The test failed because the record  didn't have the right workflow instance key. I don't recall right now whether the error was in the assertion or the error occurred while running the tests and executing the logic. If you want to know more, I could remove the line again and see which test fails.

Reply:
Since I don't understand why, I feel that it hides a potential bug. Please find out why the workflowinstance key isn't written on the record

Reply:
Not sure exactly what you mean by "Please find out why the workflowinstance key isn't written on the record". How can I find out why something is not being done???

Anyways, I removed this line just to see whether the test might now pass due to other changes we made. It still fails. It fails in `JobIncidentTest.shouldCreateIncidentWithJobErrorMessage()`

This test writes a couple of commands using these lines:
```
// when
    ENGINE
        .job()
        .ofInstance(workflowInstanceKey)
        .withType(JOB_TYPE)
        .withErrorMessage("failed job")
        .fail();
```

And then looks for a record with those lines:
```
final Record failCommand =
        RecordingExporter.jobRecords()
            .withIntent(JobIntent.FAIL)
            .withWorkflowInstanceKey(workflowInstanceKey)
            .getFirst();
```

Which is where the test fails. There is a record with that intent, but it has -1 as workflow instance key.

The `ofInstance(..)` method is where I made the change. I don't know why it didn't set the workflow instance key before. To me it felt like an oversight. I don't see how consistent records can be written this way. So honestly, I didn't think much of it when I made the change. And I still see the change as an improvement.

But your question made me curious why it had worked in the past. And there was indeed a subtle difference:

```
    final Record failedEvent =
        RecordingExporter.jobRecords()
            .withIntent(JobIntent.FAILED)
            .withWorkflowInstanceKey(workflowInstanceKey)
            .getFirst();
```

So here it is looking for the event instead of the command. And processing being the way it is the event has the correct workflow instance key. 

So now when I undo my change in this class and I undo my change in `JobClient` the test passes again. I can only assume that it may have had to do with the problem of events not being processed if there was no command processor registered for the event.

So to take stock, I could undo both changes and then have a smaller change set.

I am a little bit stubborn here, though. I like the test based on the command better. I find it clearer. The way this is used later on is in this assertion:

```
    assertThat(incidentCommand.getSourceRecordPosition())
        .isEqualTo(failedEvent.getSourceRecordPosition());
```

Before the change it read:
```
assertThat(incidentCommand.getSourceRecordPosition()).isEqualTo(failedEvent.getPosition())
```

And if I would look for the command then it would read:
```
assertThat(incidentCommand.getSourceRecordPosition()).isEqualTo(failCommand.getPosition());
``` 

Which in my mind is easier to understand than `failedEvent.getSourceRecordPosition()`. And I also feel it is a more faithful adaptation of the test. The test checks - among others - that the incident has as source position the record that produced the incident. This used to be the FAILED event, and it now is the FAIL command. 

And I also like the change to `JobClient`. I find it immensely frustrating that these utility classes produce only half-consistent records. The job client already stores the workflow instance key internally and uses it in the `findJobkey()` method, but never uses it in writing any records. And all the other methods `withType(), withVariables(), withRetries()` they all modify the record. So this still feels like a simple oversight and an obvious improvement to make. 

So in summary. I would like to keep it as it is.

What do you think?

Reply:
I had a chat with Phil and he explained to me the purpose of the `JobClient` class. With that information I will undo my changes